snuba/utils/streams/synchronized.py

```diff
@@ -0,0 +1,203 @@
+                continue
+
+            if message is None:
+                continue
+
+            commit = message.payload
+            if commit.group not in self.__commit_log_groups:
+                continue
+
+            with self.__remote_offsets.get() as remote_offsets:
+                # NOTE: This will store data about partitions that are not
+                # actually part of the subscription or assignment. This
+                # approach (potentially) requires more memory and locking
+                # overhead (due to writing state for partitions that are not
+                # subscribed or assigned), but amortizes the cost of the
+                # initial load of the topic and makes the implementation
+                # quite a bit simpler.
+                remote_offsets[commit.group][commit.partition] = commit.offset
+
+        self.__commit_log_consumer.close()
+
+    def __check_commit_log_worker_running(self) -> None:
+        if not self.closed and not self.__commit_log_worker.running():
+            try:
+                self.__commit_log_worker.result()
+            except Exception as e:
+                raise RuntimeError("commit log consumer thread crashed") from e
+            else:
+                raise RuntimeError("commit log consumer thread unexpectedly exited")
+
+    def subscribe(
+        self,
+        topics: Sequence[Topic],
+        on_assign: Optional[Callable[[Mapping[Partition, int]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[Partition]], None]] = None,
+    ) -> None:
+        def assignment_callback(offsets: Mapping[Partition, int]) -> None:
+            for partition in offsets:
+                self.__paused.discard(partition)
+
+            if on_assign is not None:
+                on_assign(offsets)
+
+        def revocation_callback(partitions: Sequence[Partition]) -> None:
+            for partition in partitions:
+                self.__paused.discard(partition)
+
+            if on_revoke is not None:
+                on_revoke(partitions)
+
+        return self.__consumer.subscribe(
+            topics, on_assign=assignment_callback, on_revoke=revocation_callback
+        )
+
+    def unsubscribe(self) -> None:
+        return self.__consumer.unsubscribe()
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
+        self.__check_commit_log_worker_running()
+
+        # Resume any partitions that can be resumed (where the local
+        # offset is less than the remote offset.)
+        resume_candidates = set(self.__consumer.paused()) - self.__paused
+        if resume_candidates:
+            local_offsets = self.tell()
+            resume_partitions = []
+
+            with self.__remote_offsets.get() as remote_offsets:
+                for partition in resume_candidates:
+                    remote_offset = min(
+                        offsets.get(partition, 0) for offsets in remote_offsets.values()
+                    )
```

Comment:
If I read this correctly this is the min of the offset on the partition across different consumer groups. Do you envision a use case for this consumer to consume the commit logs on multiple groups (other than running the migration needed to cut over from the events table to the errors table). 
We will need to be careful when using this feature since it strongly couples the functioning of multlitple consumers that work on different groups. If one stops for any reason, this synchronized consumer does not make any progress for all of them, until we do not either restart the stopped consumer or we restart this one by dropping the group that is not making progress.

Maybe a warning on the constructor close to the `commit_log_groups` parameter that explains that if one group stops this makes no progress and how to recover may be helpful.

Reply:
> If I read this correctly this is the min of the offset on the partition across different consumer groups. Do you envision a use case for this consumer to consume the commit logs on multiple groups (other than running the migration needed to cut over from the events table to the errors table).

I remember we discussed one months ago, but I don't recall specifically what it was. My guess is that it was probably along the lines of what you are talking about — migrating from one table schema to another by running another consumer on the same input set.

> Maybe a warning on the constructor close to the commit_log_groups parameter that explains that if one group stops this makes no progress and how to recover may be helpful.

Yeah, that seems like a good idea. I'll add that to the documentation. (Read: I will write documentation for this, since there is none right now 😳 and make sure that is included.)

Reply:
Added this so we don't forget how this works: a1b4e13bb5a17aa82a75ffadf382c4b9fc79c5b0